libsolidity/codegen/mlir/Passes.cpp

```diff
@@ -194,12 +199,38 @@ bool solidity::mlirgen::doJob(JobSpec const &job, mlir::MLIRContext &ctx,
 
     // Set up and run the asm printer
     llvm::legacy::PassManager llvmPassMgr;
-    tgtMach->addPassesToEmitFile(llvmPassMgr, llvm::outs(),
+
+    llvm::SmallString<0> outStreamData;
+    llvm::raw_svector_ostream outStream(outStreamData);
+    tgtMach->addPassesToEmitFile(llvmPassMgr, outStream,
                                  /*DwoOut=*/nullptr,
                                  job.action == Action::PrintAsm
                                      ? llvm::CodeGenFileType::CGFT_AssemblyFile
                                      : llvm::CodeGenFileType::CGFT_ObjectFile);
     llvmPassMgr.run(*llvmMod);
+
+    if (job.action == Action::PrintAsm) {
+      llvm::outs() << outStream.str();
+
+      // Print the bytecode in hex.
+    } else {
+      auto obj = llvm::MemoryBuffer::getMemBuffer(
+          /*InputData=*/outStream.str(), /*BufferName=*/"",
+          /*RequiresNullTerminator=*/false);
+      LLVMMemoryBufferRef cBindingBytecode = nullptr;
+      char *errMsg = nullptr;
+      if (LLVMLinkEraVM(llvm::wrap(obj.get()), &cBindingBytecode,
+                        /*metadataPtr=*/nullptr, /*metadataSize=*/0, &errMsg)) {
+        llvm_unreachable(errMsg);
```

Comment:
Please, also note that theoretically the `errMsg` should be disposed, as `LLVMLinkEraVM` allocates a memory buffer for the massage. For this you can use `LLVMDisposeMessage(errMsg)`. But as long as `llvm_unreachable` is called here, this has no meaning.